casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -500,74 +500,95 @@ object Validate {
-                   Applicative[F].pure(Right(Valid))
-               }
-    } yield result
-  }
+  ): F[Either[InvalidBlock, ValidBlock]] =
+    dag.latestMessage(b.sender).flatMap {
+      case Some(latestMessage) =>
+        val latestMessagesOfBlock = ProtoUtil.toLatestMessageHashes(b.justifications)
+        val latestMessagesFromSenderView =
+          ProtoUtil.toLatestMessageHashes(latestMessage.justifications)
+        justificationRegressionsGivenLatestMessages[F](
+          b,
+          dag,
+          latestMessagesOfBlock,
+          latestMessagesFromSenderView,
+          genesis
+        )
+      case None =>
+        // We cannot have a justification regression if we don't have a previous latest message from sender
+        Applicative[F].pure(Right(Valid))
+    }
 
-  private def justificationRegressionsAux[F[_]: Monad: Log: BlockStore](
+  private def justificationRegressionsGivenLatestMessages[F[_]: Sync: Log](
       b: BlockMessage,
-      latestMessagesOfBlock: Map[Validator, BlockHash],
-      latestMessagesFromSenderView: Map[Validator, BlockHash],
+      dag: BlockDagRepresentation[F],
+      currentLatestMessages: Map[Validator, BlockHash],
+      previousLatestMessages: Map[Validator, BlockHash],
       genesis: BlockMessage
   ): F[Either[InvalidBlock, ValidBlock]] =
-    for {
-      containsJustificationRegression <- latestMessagesOfBlock.toList.existsM {
-                                          case (validator, currentBlockJustificationHash) =>
-                                            if (validator == b.sender) {
-                                              // We let checkEquivocations handle this case
-                                              false.pure[F]
-                                            } else {
-                                              val previousBlockJustificationHash =
-                                                latestMessagesFromSenderView.getOrElse(
-                                                  validator,
-                                                  genesis.blockHash
-                                                )
-                                              isJustificationRegression[F](
-                                                currentBlockJustificationHash,
-                                                previousBlockJustificationHash
-                                              )
-                                            }
-                                        }
-      status <- if (containsJustificationRegression) {
-                 for {
-                   _ <- Log[F].warn(ignore(b, "block contains justification regressions."))
-                 } yield Left(JustificationRegression)
-               } else {
-                 Applicative[F].pure(Right(Valid))
-               }
-    } yield status
+    currentLatestMessages.toList.tailRecM {
```

Comment:
this is very hard to read

Reply:
Hmm that is unfortunate. I tried really hard to somehow not duplicate `val previousBlockJustificationHash` so that I can include the value in the log message. I suppose I should revert this back and just duplicate the value?